File: SOLID Principles/C#/OCP.cs
// O - Open/Closed Principle (OCP)
// Bad Implementation Example:
using System;
namespace OCP
{

  public class PayProcessor
  {
    public void Process(string PaymentMethod, int amount)
    {
      switch (PaymentMethod)
      {
        case "Credit Card":
          Console.WriteLine($"Paid {amount} using Credit Card");
          break;
        case "Apple Pay":
          Console.WriteLine($"Paid {amount} using Apple Pay");
          break;
        case "Paypal":
          Console.WriteLine($"Paid {amount} using Paypal");
          break;
        default:
          Console.WriteLine("Invalid Payment Method");
          break;
      }
    }
  }

  // Actual Implementation Example:

  interface IPaymentMethod
  {
    void Pay(int amount);
  }

  class CreditCardPayment : IPaymentMethod
  {
    public void Pay(int amount)
    {
      Console.WriteLine($"Paid {amount} using Credit Card");
    }
  }

  class PayPalPayment : IPaymentMethod
  {
    public void Pay(int amount)
    {
      Console.WriteLine($"Paid {amount} using PayPal");
    }
  }

  class CryptoPayment : IPaymentMethod
  {
    public void Pay(int amount)
    {

      Console.WriteLine($"Paid {amount} using Crypto");
    }
  }

  class PaymentProcessor
  {
    public void Process(IPaymentMethod PaymentMethod, int amount)
    {
      PaymentMethod.Pay(amount);
    }
  }

  //  Modified: Extension 
  class ApplePayPayment : IPaymentMethod
  {
    public void Pay(int amount)
    {
      Console.WriteLine($"Paid {amount} using Apple Pay");
    }
  }


  class Program
  {
    static void Main(string[] args)
    {
      var processor = new PaymentProcessor();

      processor.Process(new CreditCardPayment(), 200);
      processor.Process(new ApplePayPayment(), 150);
      processor.Process(new CryptoPayment(), 500);
    }
  }

}

File: SOLID Principles/Typescript/OCP.ts
// O - Open/Closed Principle (OCP)
// Bad Implementation Example:
class PayProcessor {
  process(paymentMethod: string, amount: number): void {
    switch (paymentMethod) {
      case "Credit Card":
        console.log(`Paid ${amount} using Credit Card`);
        break;
      case "Apple Pay":
        console.log(`Paid ${amount} using Apple Pay`);
        break;
      case "Paypal":
        console.log(`Paid ${amount} using Paypal`);
        break;
      default:
        console.log(`Invalid Payment Method`);
    }
  }
}

// Example usage:
const BadProcessor = new PayProcessor();
BadProcessor.process("Credit Card", 100); // Paid 100 using Credit Card
BadProcessor.process("Apple Pay", 50);    // Paid 50 using Apple Pay
BadProcessor.process("Bitcoin", 200);     // Invalid Payment Method


// Actual Implementation Example:

interface IPaymentMethod {
  pay(amount: number): void;
}

class CreditCardPayment implements IPaymentMethod {
  pay(amount: number): void {
    console.log(`Paid ${amount} using Credit Card`);
  }
}

class PayPalPayment implements IPaymentMethod {
  pay(amount: number): void {
    console.log(`Paid ${amount} using PayPal`);
  }
}

class CryptoPayment implements IPaymentMethod {
  pay(amount: number): void {
    console.log(`Paid ${amount} using Crypto`);
  }
}

class PaymentProcessor {
  process(paymentMethod: IPaymentMethod, amount: number): void {
    paymentMethod.pay(amount);
  }
}

//  Modified: Extension 
class ApplePayPayment implements IPaymentMethod {
  pay(amount: number): void {
    console.log(`Paid ${amount} using Apple Pay`);
  }
}




const processor = new PaymentProcessor();
processor.process(new CreditCardPayment(), 200);
processor.process(new ApplePayPayment(), 200);





